src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -114,85 +116,94 @@ private bool IsLocallySuppressed (int id, ICustomAttributeProvider provider, out
+			switch (provider.MetadataToken.TokenType) {
+			case TokenType.Module:
+				module = provider as ModuleDefinition;
+				break;
+			case TokenType.Assembly:
+				module = (provider as AssemblyDefinition).MainModule;
+				break;
+			case TokenType.TypeDef:
+				module = (provider as TypeDefinition).Module;
+				break;
+			case TokenType.Method:
+			case TokenType.Property:
+			case TokenType.Field:
+			case TokenType.Event:
+				module = (provider as IMemberDefinition).DeclaringType.Module;
+				break;
+			default:
+				_context.LogMessage (MessageContainer.CreateInfoMessage ("`UnconditionalSuppressMessage` attribute was placed in an language element which is currently not supported."));
+				return;
+			}
 
-				_globalSuppressions = suppressions;
+			Debug.Assert (module != null);
+			AssemblyDefinition assembly = module.Assembly;
+			if (!_globalSuppressions.InitializedAssemblies.Contains (assembly)) {
+				DecodeGlobalSuppressMessageAttributes (assembly);
+				_globalSuppressions.InitializedAssemblies.Add (assembly);
 			}
 		}
 
-		private void DecodeGlobalSuppressMessageAttributes (ICustomAttributeProvider provider, GlobalSuppressions globalSuppressions)
+		private void DecodeGlobalSuppressMessageAttributes (AssemblyDefinition assembly)
 		{
-			var attributes = provider.CustomAttributes.
-				Where (a => a.AttributeType.Name == "UnconditionalSuppressMessageAttribute" && a.AttributeType.Namespace == "System.Diagnostics.CodeAnalysis");
-			foreach (var instance in attributes) {
-				SuppressMessageInfo info;
-				if (!TryDecodeSuppressMessageAttributeData (instance, out info))
-					continue;
-
-				ModuleDefinition module = null;
-				switch (provider.MetadataToken.TokenType) {
-				case TokenType.Module:
-					module = provider as ModuleDefinition;
-					break;
-				case TokenType.Assembly:
-					module = (provider as AssemblyDefinition).MainModule;
-					break;
-				case TokenType.TypeDef:
-					module = (provider as TypeDefinition).Module;
-					break;
-				case TokenType.Method:
-				case TokenType.Property:
-				case TokenType.Field:
-				case TokenType.Event:
-					module = (provider as IMemberDefinition).DeclaringType.Module;
-					break;
-				default:
-					_context.LogMessage (MessageContainer.CreateInfoMessage ("`UnconditionalSuppressMessage` attribute was placed in an language element which is currently not supported."));
-					continue;
-				}
-
-				Debug.Assert (module != null);
-				if (info.Target == null && (info.Scope == "module" || info.Scope == null)) {
-					AddLocalSuppression (instance, module);
-					continue;
-				}
+			void LookForSuppressions (ICustomAttributeProvider provider) {
+				ModuleDefinition module = provider is ModuleDefinition ? provider as ModuleDefinition : (provider as AssemblyDefinition).MainModule; 
+				var attributes = provider.CustomAttributes.
+					Where (a => a.AttributeType.Name == "UnconditionalSuppressMessageAttribute" && a.AttributeType.Namespace == "System.Diagnostics.CodeAnalysis");
```

Comment:
Extract the check to static method. I think you are using it at multiple places already